libmamba/src/core/pool.cpp

```diff
@@ -21,119 +19,146 @@ extern "C"
+            {
+                dbg->first->info(log);
+            }
+            dbg->second.clear();
         }
 
-        auto log = Console::hide_secrets(dbg->second);
-        if (type & SOLV_FATAL || type & SOLV_ERROR)
-        {
-            dbg->first->error(log);
-        }
-        else if (type & SOLV_WARN)
+        void libsolv_delete_pool(::Pool* pool)
         {
-            dbg->first->warn(log);
+            LOG_INFO << "Freeing pool.";
+            pool_free(pool);
         }
-        else if (Context::instance().verbosity > 2)
+
+    }
+
+    struct MPool::MPoolData
+    {
+        MPoolData()
+            : pool(pool_create(), &libsolv_delete_pool)
         {
-            dbg->first->info(log);
         }
-        dbg->second.clear();
-    }
+
+        std::unique_ptr<Pool, decltype(&libsolv_delete_pool)> pool;
+        std::pair<spdlog::logger*, std::string> debug_logger = {};
+        std::list<MRepo> repo_list = {};
+    };
 
     MPool::MPool()
+        : m_data(std::make_shared<MPoolData>())
     {
-        m_pool = pool_create();
-        pool_setdisttype(m_pool, DISTTYPE_CONDA);
+        pool_setdisttype(pool(), DISTTYPE_CONDA);
         set_debuglevel();
     }
 
-    MPool::~MPool()
+    MPool::~MPool() = default;
+
+    Pool* MPool::pool()
+    {
+        return m_data->pool.get();
+    }
+
+    Pool const* MPool::pool() const
     {
-        LOG_INFO << "Freeing pool.";
-        m_repo_list.clear();
-        pool_free(m_pool);
+        return m_data->pool.get();
     }
 
     void MPool::set_debuglevel()
     {
         // ensure that debug logging goes to stderr as to not interfere with stdout json output
-        m_pool->debugmask |= SOLV_DEBUG_TO_STDERR;
+        pool()->debugmask |= SOLV_DEBUG_TO_STDERR;
         if (Context::instance().verbosity > 2)
         {
-            pool_setdebuglevel(m_pool, Context::instance().verbosity - 1);
+            pool_setdebuglevel(pool(), Context::instance().verbosity - 1);
             auto logger = spdlog::get("libsolv");
-            m_debug_logger.first = logger.get();
-            pool_setdebugcallback(m_pool, &libsolv_debug_callback, &m_debug_logger);
+            m_data->debug_logger.first = logger.get();
+            pool_setdebugcallback(pool(), &libsolv_debug_callback, &(m_data->debug_logger));
         }
     }
 
     void MPool::create_whatprovides()
     {
-        pool_createwhatprovides(m_pool);
+        pool_createwhatprovides(pool());
     }
 
     MPool::operator Pool*()
     {
-        return m_pool;
+        return pool();
     }
 
     MPool::operator Pool const*() const
     {
-        return m_pool;
+        return pool();
     }
 
     std::vector<Id> MPool::select_solvables(Id matchspec, bool sorted) const
     {
         MQueue job, solvables;
         job.push(SOLVER_SOLVABLE_PROVIDES, matchspec);
-        selection_solvables(m_pool, job, solvables);
+        selection_solvables(const_cast<Pool*>(pool()), job, solvables);
```

Comment:
The method was const (the `Pool` pointer was a `Pool* const`) but the libsolv function is not `const` (but libsolv does not seem fully const correct).